File: backend/src/notifications/teams.service.ts
import { Injectable, Logger, Inject, forwardRef } from '@nestjs/common';
import { SettingsService } from '../settings/settings.service';

@Injectable()
export class TeamsService {
  private readonly logger = new Logger(TeamsService.name);

  constructor(
    @Inject(forwardRef(() => SettingsService))
    private settingsService: SettingsService
  ) {}

  async sendVulnerabilityAlert(scanId: string, projectName: string, summary: any) {
    const settings = await this.settingsService.getTeamsSettings();
    if (!settings.enabled || !settings.webhookUrl || !this.isValidWebhookUrl(settings.webhookUrl)) {
      this.logger.warn(`Teams notification disabled or webhook URL (${settings.webhookUrl}) is invalid/unsafe.`);
      return;
    }

    this.logger.log(`Sending Teams notification for scan ${scanId}`);

    const payload = {
      type: "message",
      attachments: [
        {
          contentType: "application/vnd.microsoft.card.adaptive",
          content: {
            type: "AdaptiveCard",
            $schema: "http://adaptivecards.io/schemas/adaptive-card.json",
            version: "1.2",
            body: [
              {
                type: "TextBlock",
                size: "Medium",
                weight: "Bolder",
                text: `🚨 Alerte de Sécurité Zanshin - ${projectName}`,
                color: "Attention"
              },
              {
                type: "TextBlock",
                text: `Des vulnérabilités ont été détectées lors du scan **${scanId}**.`,
                wrap: true
              },
              {
                type: "FactSet",
                facts: [
                  { title: "Critique", value: `${summary.critical}` },
                  { title: "Élevé", value: `${summary.high}` },
                  { title: "Moyen", value: `${summary.medium}` },
                  { title: "Faible", value: `${summary.low}` }
                ]
              }
            ],
            actions: [
              {
                type: "Action.OpenUrl",
                title: "Voir le rapport",
                url: "http://localhost:4200" // Idéalement l'URL de votre instance
              }
            ]
          }
        }
      ]
    };

    try {
      const response = await fetch(settings.webhookUrl, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload)
      });

      if (!response.ok) {
        const errorText = await response.text();
        throw new Error(`Teams Webhook failed: ${response.status} ${errorText}`);
      }
      
      this.logger.log('Teams notification sent successfully');
    } catch (error) {
      this.logger.error(`Failed to send Teams notification: ${error.message}`);
    }
  }

  async sendTestMessage(webhookUrl: string) {
    if (!this.isValidWebhookUrl(webhookUrl)) {
      throw new Error('Invalid or unsafe webhook URL provided.');
    }

    const payload = {
      type: "message",
// ... rest of the code unchanged
      attachments: [
        {
          contentType: "application/vnd.microsoft.card.adaptive",
          content: {
            type: "AdaptiveCard",
            $schema: "http://adaptivecards.io/schemas/adaptive-card.json",
            version: "1.2",
            body: [
              {
                type: "TextBlock",
                size: "Large",
                weight: "Bolder",
                text: "Zanshin - Test de Connexion Teams"
              },
              {
                type: "TextBlock",
                text: "Félicitations ! Votre webhook Teams est correctement configuré.",
                wrap: true
              }
            ]
          }
        }
      ]
    };

    const response = await fetch(webhookUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload)
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`Erreur Teams (${response.status}): ${errorText}`);
    }

    return { message: 'Message de test envoyé' };
  }
}

File: backend-java/src/main/java/com/asmolabs/zanshin/notifications/services/TeamsService.java
package com.asmolabs.zanshin.notifications.services;

import com.asmolabs.zanshin.settings.services.SettingsService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.annotation.Lazy;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClient;

import java.net.URI;
import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
@Slf4j
public class TeamsService {

    @Lazy
    private final SettingsService settingsService;
    private final RestClient restClient = RestClient.create();

    public void sendVulnerabilityAlert(Long scanId, String projectName, Map<String, Object> summary) {
        Map<String, Object> settings = settingsService.getTeamsSettings();
        Boolean enabled = (Boolean) settings.get("enabled");
        String webhookUrl = (String) settings.get("webhookUrl");

        if (enabled == null || !enabled || webhookUrl == null || !isValidWebhookUrl(webhookUrl)) {
            log.warn("Teams notification disabled or webhook URL is invalid/unsafe.");
            return;
        }

        log.info("Sending Teams notification for scan {}", scanId);

        Map<String, Object> payload = Map.of(
            "type", "message",
            "attachments", List.of(Map.of(
                "contentType", "application/vnd.microsoft.card.adaptive",
                "content", Map.of(
                    "type", "AdaptiveCard",
                    "$schema", "http://adaptivecards.io/schemas/adaptive-card.json",
                    "version", "1.2",
                    "body", List.of(
                        Map.of(
                            "type", "TextBlock",
                            "size", "Medium",
                            "weight", "Bolder",
                            "text", "🚨 Alerte de Sécurité Zanshin - " + projectName,
                            "color", "Attention"
                        ),
                        Map.of(
                            "type", "TextBlock",
                            "text", "Des vulnérabilités ont été détectées lors du scan **" + scanId + "**.",
                            "wrap", true
                        ),
                        Map.of(
                            "type", "FactSet",
                            "facts", List.of(
                                Map.of("title", "Critique", "value", String.valueOf(summary.getOrDefault("critical", 0))),
                                Map.of("title", "Élevé", "value", String.valueOf(summary.getOrDefault("high", 0))),
                                Map.of("title", "Moyen", "value", String.valueOf(summary.getOrDefault("medium", 0))),
                                Map.of("title", "Faible", "value", String.valueOf(summary.getOrDefault("low", 0)))
                            )
                        )
                    ),
                    "actions", List.of(Map.of(
                        "type", "Action.OpenUrl",
                        "title", "Voir le rapport",
                        "url", "http://localhost:4200"
                    ))
                )
            ))
        );

        try {
            restClient.post()
                    .uri(webhookUrl)
                    .body(payload)
                    .retrieve()
                    .toBodilessEntity();
            log.info("Teams notification sent successfully");
        } catch (Exception e) {
            log.error("Failed to send Teams notification: {}", e.getMessage());
        }
    }

    public Map<String, Object> sendTestMessage(String webhookUrl) {
        if (!isValidWebhookUrl(webhookUrl)) {
            throw new RuntimeException("Invalid or unsafe webhook URL provided.");
        }

        Map<String, Object> payload = Map.of(
            "type", "message",
            "attachments", List.of(Map.of(
                "contentType", "application/vnd.microsoft.card.adaptive",
                "content", Map.of(
                    "type", "AdaptiveCard",
                    "$schema", "http://adaptivecards.io/schemas/adaptive-card.json",
                    "version", "1.2",
                    "body", List.of(
                        Map.of(
                            "type", "TextBlock",
                            "size", "Large",
                            "weight", "Bolder",
                            "text", "Zanshin - Test de Connexion Teams"
                        ),
                        Map.of(
                            "type", "TextBlock",
                            "text", "Félicitations ! Votre webhook Teams est correctement configuré.",
                            "wrap", true
                        )
                    )
                )
            ))
        );

        try {
            restClient.post()
                    .uri(webhookUrl)
                    .body(payload)
                    .retrieve()
                    .toBodilessEntity();
            return Map.of("message", "Message de test envoyé");
        } catch (Exception e) {
            log.error("Teams Test failed: {}", e.getMessage());
            throw new RuntimeException("Erreur Teams : " + e.getMessage());
        }
    }

    private boolean isValidWebhookUrl(String url) {
        if (url == null || url.isEmpty()) return false;
        try {
            URI uri = new URI(url);
            String host = uri.getHost();
            if (host == null) return false;
            
            List<String> allowedDomains = List.of(
                "outlook.office.com",
                "outlook.office365.com",
                "webhook.office.com"
            );
            
            boolean domainAllowed = allowedDomains.stream().anyMatch(host::endsWith);
            return domainAllowed && "https".equalsIgnoreCase(uri.getScheme());
        } catch (Exception e) {
            return false;
        }
    }
}
